elastic4s-core/src/main/scala/com/sksamuel/elastic4s/Reader.scala

```diff
@@ -7,4 +12,94 @@ trait Reader[-U] {
 
 trait HitAs[T] {
   def as(hit: RichSearchHit): T
-}
\ No newline at end of file
+}
+
+trait HitRead[T] {
+  def as(hit: RichSearchHitLike): T Or Every[ErrorMessage]
+}
+
+trait HitFieldRead[T] {
+  def as(prefix:String)(hit: RichSearchHitField): T Or Every[ErrorMessage]
+  def as(hit: RichSearchHitField): T Or Every[ErrorMessage]=as("")(hit)
+}
+
+trait LPHitFieldReads{
+  implicit def hitRead2HitFieldRead[T](implicit hitRead:HitRead[T]) =new HitFieldRead[T] {
+    import scala.collection.JavaConverters._
+    override def as(prefix: String)(hit: RichSearchHitField): Or[T, Every[ErrorMessage]] = hit match {
+      case MissingRichSearchField(name)=> Bad(One(s"$prefix $name field is missing"))
+      case field=> field.value[Any] match {
+        case map:java.util.Map[_,_] => hitRead.as(RichMapSearchHit(map.asInstanceOf[java.util.Map[String,Any]].asScala.toMap))
+        case other => Bad(One(s"$prefix ${hit.name} expected a object (Map[String,Any]) got $other"))
```

Comment:
Hello,

I won't be able to update the PR until September but I'll do it if it is
still there when I get back.

Cheers
Jean

Le jeu. 6 août 2015 22:20, Martin Grigorov <notifications@github.com> a
écrit :

> In elastic4s-core/src/main/scala/com/sksamuel/elastic4s/Reader.scala
> <https://github.com/sksamuel/elastic4s/pull/353#discussion_r36460539>:
>
> > +  def as(hit: RichSearchHitLike): T Or Every[ErrorMessage]
> > +}
> > +
> > +trait HitFieldRead[T] {
> > +  def as(prefix:String)(hit: RichSearchHitField): T Or Every[ErrorMessage]
> > +  def as(hit: RichSearchHitField): T Or Every[ErrorMessage]=as("")(hit)
> > +}
> > +
> > +trait LPHitFieldReads{
> > +  implicit def hitRead2HitFieldRead[T](implicit hitRead:HitRead[T]) =new HitFieldRead[T] {
> > +    import scala.collection.
>
> JavaConverters._
> > +    override def as(prefix: String)(hit: RichSearchHitField): Or[T, Every[ErrorMessage]] = hit match {
> > +      case MissingRichSearchField(name)=> Bad(One(s"$prefix $name field is missing"))
> > +      case field=> field.value[Any] match {
> > +        case map:java.util.Map[_,_] => hitRead.as(RichMapSearchHit(map.asInstanceOf[java.util.Map[String,Any]].asScala.toMap))
> > +        case other => Bad(One(s"$prefix ${hit.name} expected a object (Map[String,Any]) got $other"))
>
>
>    1. aN object
>    2. put a comma before got. I think it will be easier to read
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sksamuel/elastic4s/pull/353/files#r36460539>.
>